pandas/core/groupby/groupby.py

```diff
@@ -2116,7 +2121,13 @@ def ngroup(self, ascending: bool = True):
         """
         with _group_selection_context(self):
             index = self._selected_obj.index
-            result = Series(self.grouper.group_info[0], index)
+
+            # GH28330 preserve subclassed Series/DataFrames
```

Comment:
can you make this more concise e.g.

klass = self.obj_constructor if isinstance(self.obj, Series) else self.obj._constructor_sliced
result = klass(.....)

Reply:
also ok would be to add a property to GroupBy which return this e.g.

```
@property
def _constructor(self):
.....
```

Reply:
Good idea -- I added a `GroupBy._constructor` property to consolidate the relevant logic in one place. This allowed me to change `ngroup`, `cumcount`, and `size` to use more concise statements.